### ISSUE TITLE
Scheduling tasks with high priority after-the-fact

### ISSUE THREAD

**Issue by baldurk** [dougbinks/enkiTS#77]
This may be out of scope for the library, or potentially I've missed a way to do it. What I would like to be able to do is add a number of tasks with no particular priority and have them be scheduled in parallel in no particular order, but I would want the ability to immediately run one if I discover that it's particularly needed soon (while allowing others to still be scheduled as/when).

The sketched idea is to have the main thread push tasks for a bunch of deferred operations as it's processing something, but then if the results of one of those deferred operations is needed it wants to either immediately run that task on the main thread or spinloop if it's already been scheduled. The tasks can come in any order so you could push 100 different tasks and then want the results from number 50 before continuing, so waiting for 1-49 would not be ideal as then the main thread can't continue processing in parallel with them.

As far as I can tell unless I'm misunderstanding the code, WaitForTask does not prioritise the task you pass in to wait on but it looks at all tasks at equal priority. I don't see a way to elevate the priority of a task after it has been created either. I also wondered if the way to solve this would be to add a new task with a dependency on the one I want to run, but at least from a layperson's eye that doesn't seem like it would change the scheduling order.

Is there a way to do this, and/or is it something you'd be interested in supporting? In principle this could be refactored such that there is no main thread needing to sync, and everything including its processing becomes task-based so this translates into dependency management, but that's further than I'd want to go.

**Comment by dougbinks** [comment]
If you can keep track of any work done you can also only process what's needed in the new task, but it might be simpler just to discard it depending on what you're doing.

**Comment by dougbinks** [comment]
I can think of a way to do this in enkiTS, with a little extra work. Elevating a tasks priority would be fairly expensive, but it could be done. The work the threads are already doing cannot currently be interrupted, so this wouldn't guarantee that it would get processed immediately but it would be able to move it to the front of all other tasks or subtasks (the internal unit of work) which had not yet started processing. I'll think a bit more before committing to writing any code and get back to you.

**Comment by baldurk** [comment]
That totally makes sense, it's co-operative rather than pre-emptive after all. I was only after an "as soon as possible" scheduling rather than immediate. For my part though I think your suggestion of cancelling and re-queuing with high priority is sufficient, so if this would be expensive or awkward to implement then feel free to close.
